<a href="https://colab.research.google.com/github/KangPixel/AIFFEL_quest_cr/blob/master/Python/Project/Node_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 6-1. digit : 손글씨를 분류해 보기.

In [57]:
import sklearn
# 필요 모듈
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
# 모델들
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 데이터 준비
digits = load_digits()

# Feature Data 지정하기
# Label Data 지정하기
# Target Names 출력해 보기
# 데이터 Describe 해 보기

digits_data = digits.data
digits_label = digits.target
print(digits.target_names)
print('-'*50)
print(f"digits_data shape: {digits_data.shape}")
print(f"digits_label shape: {digits_label.shape}")
print(f"First feature sample:\n{digits_data[0]}")
print(f"First label sample: {digits_label[0]}")
# print(digits.DESCR)

# train, test데이터 분리
x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_label,
                                                    test_size=0.2,
                                                    random_state=7)

[0 1 2 3 4 5 6 7 8 9]
--------------------------------------------------
digits_data shape: (1797, 64)
digits_label shape: (1797,)
First feature sample:
[ 0.  0.  5. 13.  9.  1.  0.  0.  0.  0. 13. 15. 10. 15.  5.  0.  0.  3.
 15.  2.  0. 11.  8.  0.  0.  4. 12.  0.  0.  8.  8.  0.  0.  5.  8.  0.
  0.  9.  8.  0.  0.  4. 11.  0.  1. 12.  7.  0.  0.  2. 14.  5. 10. 12.
  0.  0.  0.  0.  6. 13. 10.  0.  0.  0.]
First label sample: 0


In [58]:
# 테스트 하는 부분 수
# digits_data.shape
x_test.shape

(360, 64)

In [59]:
# 다양한 모델 학습
# Decision Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(x_train, y_train)
y_dt_pred = decision_tree.predict(x_test)

print(classification_report(y_test, y_dt_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



In [60]:
# Random Forest
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(x_train, y_train)
y_rf_pred = random_forest.predict(x_test)

print(classification_report(y_test, y_rf_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



In [61]:
# SVM
svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_svm_pred = svm_model.predict(x_test)

print(classification_report(y_test, y_svm_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [62]:
# SGD Classifier
sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_sgd_pred = sgd_model.predict(x_test)

print(classification_report(y_test, y_sgd_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.90      0.86      0.88        42
           2       0.95      1.00      0.98        40
           3       0.94      0.88      0.91        34
           4       1.00      0.97      0.99        37
           5       0.96      0.96      0.96        28
           6       0.96      0.93      0.95        28
           7       0.97      0.94      0.95        33
           8       0.91      0.91      0.91        43
           9       0.81      0.94      0.87        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360



In [63]:
# Logistic Regression
logistic_model = LogisticRegression()
logistic_model.fit(x_train, y_train)
y_lg_pred = logistic_model.predict(x_test)

print(classification_report(y_test, y_lg_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## 모델 평가
### 종합적인 F1-score를 기준

Decision Tree  
- 평균 86%  

Random Forest  
- 평균 96%  

SVM
- 평균 99%

SGD Classifier
- 평균 94%

Logistic Regression
- 평균 95%
---
SVM이 정확도가 가장 높게 학습함.
---

# 6-2. wine : 와인을 분류해 보기.

In [64]:
# 필요 모듈
from sklearn.datasets import load_wine
# 이미 위에서 import 했음.
# 만약 새로 할 경우 아래 주석을 풀고 실행
'''
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
# 모델들
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
'''

# 데이터 준비
wine = load_wine()

# Feature Data 지정하기
# Label Data 지정하기
# Target Names 출력해 보기
# 데이터 Describe 해 보기
wine_data = wine.data
wine_label = wine.target
print(wine.target_names)
print('-'*50)
print(f"digits_data shape: {wine_data.shape}")
print(f"digits_label shape: {wine_label.shape}")
print(f"First feature sample:\n{wine_data[0]}")
print(f"First label sample: {wine_label[0]}")

# train, test 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.2,
                                                    random_state=7)

['class_0' 'class_1' 'class_2']
--------------------------------------------------
digits_data shape: (178, 13)
digits_label shape: (178,)
First feature sample:
[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
 2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
First label sample: 0


In [65]:
# 다양한 모델로 학습
# 6-1에서 해본 결과 함수 불러오는 부분은 생략 가능
# 6-1에서 해본 결과 학습 부분만 따로 모아야 가시성이 좋아보임
# 만약 새로 할 경우 아래 주석을 풀고 실행

# Decision Tree
# decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(x_train, y_train)
y_dt_pred = decision_tree.predict(x_test)

# Random Forest
# random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(x_train, y_train)
y_rf_pred = random_forest.predict(x_test)

# SVM
# svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_svm_pred = svm_model.predict(x_test)

# SGD Classifier
# sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_sgd_pred = sgd_model.predict(x_test)

# Logistic Regression
# logistic_model = Logistic
logistic_model.fit(x_train, y_train)
y_lg_pred = logistic_model.predict(x_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [66]:
# 모델 평가
print('Decision Tree')          # Decision Tree 결과
print(classification_report(y_test, y_dt_pred))
print('-'*50)

print('Random Forest')          # Random Forest 결과
print(classification_report(y_test, y_rf_pred))
print('-'*50)

print('SVM')                    #SVM 결과
print(classification_report(y_test, y_svm_pred))
print('-'*50)

print('SGD Classifier')         #SGD Classifier 결과
print(classification_report(y_test, y_sgd_pred))
print('-'*50)

print('Logistic Regression')    # Logistic Regression 결과
print(classification_report(y_test, y_lg_pred))

Decision Tree
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36

--------------------------------------------------
Random Forest
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

--------------------------------------------------
SVM
              precision    recall  f1-score   support

           0       0.86      0.86      0.86      

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 모델 평가
### 종합적인 F1-score를 기준

Decision Tree  
- 평균 94%  

Random Forest  
- 평균 100%  

SVM
- 평균 61%

SGD Classifier
- 평균 61%

Logistic Regression
- 평균 97%
---
Random Forest가 정확도가 가장 높게 학습함.
---

# 6-3. breast_cancer : 유방암 여부를 진단해 보기.

In [67]:
# 필요 모듈
from sklearn.datasets import load_breast_cancer
# 이미 '6-1'에서 import 했음.
# 만약 새로 할 경우 아래 주석을 풀고 실행
'''
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
# 모델들
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
'''

# 데이터 준비
cancer = load_breast_cancer()

# Feature Data 지정하기
# Label Data 지정하기
# Target Names 출력해 보기
# 데이터 Describe 해 보기
cancer_data = cancer.data
cancer_label = cancer.target
print(cancer.target_names)
print('-'*50)
print(f"digits_data shape: {cancer_data.shape}")
print(f"digits_label shape: {cancer_label.shape}")
print(f"First feature sample:\n{cancer_data[0]}")
print(f"First label sample: {cancer_label[0]}")

# train, test 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(cancer_data,
                                                    cancer_label,
                                                    test_size=0.2,
                                                    random_state=7)

['malignant' 'benign']
--------------------------------------------------
digits_data shape: (569, 30)
digits_label shape: (569,)
First feature sample:
[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]
First label sample: 0


In [68]:
# 다양한 모델로 학습
# '6-1'에서 해본 결과 함수 불러오는 부분은 생략 가능
# 만약 새로 할 경우 아래 주석을 풀고 실행

# Decision Tree
# decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(x_train, y_train)
y_dt_pred = decision_tree.predict(x_test)

# Random Forest
# random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(x_train, y_train)
y_rf_pred = random_forest.predict(x_test)

# SVM
# svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_svm_pred = svm_model.predict(x_test)

# SGD Classifier
# sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_sgd_pred = sgd_model.predict(x_test)

# Logistic Regression
# logistic_model = Logistic
logistic_model.fit(x_train, y_train)
y_lg_pred = logistic_model.predict(x_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [69]:
# 모델 평가
print('Decision Tree')          # Decision Tree 결과
print(classification_report(y_test, y_dt_pred))
print('-'*50)

print('Random Forest')          # Random Forest 결과
print(classification_report(y_test, y_rf_pred))
print('-'*50)

print('SVM')                    #SVM 결과
print(classification_report(y_test, y_svm_pred))
print('-'*50)

print('SGD Classifier')         #SGD Classifier 결과
print(classification_report(y_test, y_sgd_pred))
print('-'*50)

print('Logistic Regression')    # Logistic Regression 결과
print(classification_report(y_test, y_lg_pred))

Decision Tree
              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114

--------------------------------------------------
Random Forest
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114

--------------------------------------------------
SVM
              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90     

## 모델 평가
### 종합적인 F1-score를 기준

Decision Tree  
- 평균 91%  

Random Forest  
- 평균 100%  

SVM
- 평균 90%

SGD Classifier
- 평균 82%

Logistic Regression
- 평균 94%
---
Random Forest가 정확도가 가장 높게 학습함.  - 의심이 들정도로 100%...-
---

# 의문점
같은 값으로 돌렸을 경우  
SGD Classifier 모델이 계속 정확도가 오름.  
- 6-1.손글씨 : 93% >> 94%
- 6-2.와인&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 50% >> 61%
- 6-3.유방암 : 52% >> 82%